kategory-core/src/main/kotlin/kategory/typeclasses/MonadFilterContinuation.kt

```diff
@@ -0,0 +1,37 @@
+package kategory
+
+import kotlin.coroutines.experimental.CoroutineContext
+import kotlin.coroutines.experimental.EmptyCoroutineContext
+import kotlin.coroutines.experimental.RestrictsSuspension
+
+@RestrictsSuspension
+open class MonadFilterContinuation<F, A>(val MF: MonadFilter<F>, override val context: CoroutineContext = EmptyCoroutineContext) :
+        MonadContinuation<F, A>(MF) {
+
+    object PredicateInterrupted : RuntimeException()
+
+    override fun resumeWithException(exception: Throwable) {
+        when (exception) {
+            is PredicateInterrupted -> returnedMonad = MF.empty()
+            else -> super.resumeWithException(exception)
+        }
+    }
+
+    /**
+     * Short circuits monadic bind if `predicate == false` return the
+     * monad `empty` value.
+     */
+    fun continueIf(predicate: Boolean): Unit {
```

Comment:
Can this be implemented as `fun bindIf(predicate: () -> Boolean): Unit { if (predicate()) pure(Unit) else MF.empty()}`. That way we avoid throwing expensive exceptions as shortcircuit.